File: SearchInsertPosition.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        if nums[0] > target:
            return 0
        elif nums[-1] < target:
            return len(nums)

        #binary search
        start = 0
        end = len(nums) - 1
        while start <= end:
            mid = (start + end) // 2
            if target == nums[mid]:
                return mid
            elif nums[mid] < target and nums[mid + 1] > target:
                return mid + 1
            elif target < nums[mid]: 
                end = mid - 1 #check the left side of the list
            else:
                start = mid + 1 #check the right side of the list

File: SearchInsertPosition.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        if (nums[0] > target){
            return 0;
        }else if(nums[nums.length-1] < target){
            return nums.length;
        }

        // Perform binary search
        int l = 0;
        int r = nums.length - 1;

        while (l <= r){
            int m = l + ((r - l) / 2); //consider overflow

            if (nums[m] == target){
                return m;
            }else if (nums[m] < target && nums[m+1] > target){
                return m+1;
            }else if (nums[m] > target){
                r = m - 1;
            }else {
                l = m + 1;
            }
        }
        return -1; //I don't think this part matters, the code above should find the actual answer
    }
}

// Time: O(log n)
// Space: O(1)
